python/tests/test_parsimony.py

```diff
@@ -217,14 +226,18 @@ def hartigan_map_mutations(tree, genotypes, alleles):
                 max_allele_count = np.max(allele_count)
                 optimal_set[u, allele_count == max_allele_count] = 1
 
-    allele_count[:] = 0
-    for v in tree.roots:
-        for j in range(num_alleles):
-            allele_count[j] += optimal_set[v, j]
-    max_allele_count = np.max(allele_count)
-    optimal_root_set = np.zeros(num_alleles, dtype=int)
-    optimal_root_set[allele_count == max_allele_count] = 1
-    ancestral_state = np.argmax(optimal_root_set)
+    if ancestral_state is None:
+        allele_count[:] = 0
+        for v in tree.roots:
+            for j in range(num_alleles):
+                allele_count[j] += optimal_set[v, j]
+        max_allele_count = np.max(allele_count)
+        optimal_root_set = np.zeros(num_alleles, dtype=int)
+        optimal_root_set[allele_count == max_allele_count] = 1
+        ancestral_state = np.argmax(optimal_root_set)
+    else:
+        if ancestral_state < 0 or ancestral_state >= len(alleles):
```

Comment:
The second clause doesn't count here now. Better to put the check further with the check around line 205 so that this logic is all in one place.